scipy removed `_lazywhere`

#### Describe the bug

scipy removed `_lazywhere` in scipy/scipy#22557 which is used in statsmodels:


https://github.com/statsmodels/statsmodels/blob/24b4938c0d1b4c2a2d9eff238cf85c0f8e5575a3/statsmodels/genmod/_tweedie_compound_poisson.py#L20

https://github.com/statsmodels/statsmodels/blob/24b4938c0d1b4c2a2d9eff238cf85c0f8e5575a3/statsmodels/distributions/discrete.py#L5


#### Code Sample, a copy-pastable example if possible

```bash
python -m pip install statsmodels
python -m pip install -i https://pypi.anaconda.org/scientific-python-nightly-wheels/simple --no-deps --pre --upgrade  scipy statsmodels

```


```python
import statsmodels.api as sm
```

```python-traceback
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mathause/.conda/envs/test/lib/python3.12/site-packages/statsmodels/api.py", line 76, in <module>
    from . import datasets, distributions, iolib, regression, robust, tools
  File "/home/mathause/.conda/envs/test/lib/python3.12/site-packages/statsmodels/distributions/__init__.py", line 7, in <module>
    from .discrete import (
  File "/home/mathause/.conda/envs/test/lib/python3.12/site-packages/statsmodels/distributions/discrete.py", line 5, in <module>
    from scipy._lib._util import _lazywhere
ImportError: cannot import name '_lazywhere' from 'scipy._lib._util' (/home/mathause/.conda/envs/test/lib/python3.12/site-packages/scipy/_lib/_util.py)
``


#### Expected Output

A clear and concise description of what you expected to happen.

#### Output of ``import statsmodels.api as sm; sm.show_versions()``

...

Many errors in pre-release testing:

```
____________________________ test_deltacov_margeff _____________________________
[gw1] linux -- Python 3.12.9 /opt/hostedtoolcache/Python/3.12.9/x64/bin/python

    def test_deltacov_margeff():
        # compare with discrete margins
>       import statsmodels.discrete.tests.test_discrete as dt

statsmodels/stats/tests/test_deltacov.py:94: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/hostedtoolcache/Python/3.12.9/x64/lib/python3.12/site-packages/_pytest/assertion/rewrite.py:186: in exec_module
    exec(co, module.__dict__)
statsmodels/discrete/tests/test_discrete.py:31: in <module>
    import statsmodels.api as sm
statsmodels/api.py:76: in <module>
    from . import datasets, distributions, iolib, regression, robust, tools
statsmodels/distributions/__init__.py:7: in <module>
    from .discrete import (
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    import numpy as np
    
    from scipy.stats import rv_discrete, poisson, nbinom
    from scipy.special import gammaln
>   from scipy._lib._util import _lazywhere
E   ImportError: cannot import name '_lazywhere' from 'scipy._lib._util' (/opt/hostedtoolcache/Python/3.12.9/x64/lib/python3.12/site-packages/scipy/_lib/_util.py)
```

--- a/statsmodels/compat/scipy.py
+++ b/statsmodels/compat/scipy.py
@@ -1,6 +1,5 @@
-from packaging.version import Version, parse
-
 import numpy as np
+from packaging.version import Version, parse
 import scipy
 
 SP_VERSION = parse(scipy.__version__)
@@ -72,3 +71,63 @@ if SP_LT_16:
     from ._scipy_multivariate_t import multivariate_t  # noqa: F401
 else:
     from scipy.stats import multivariate_t  # noqa: F401
+
+
+def apply_where(  # type: ignore[explicit-any] # numpydoc ignore=PR01,PR02
+    cond, args, f1, f2=None, /, *, fill_value=None
+):
+    """
+    Run one of two elementwise functions depending on a condition.
+
+    Equivalent to ``f1(*args) if cond else fill_value`` performed elementwise
+    when `fill_value` is defined, otherwise to ``f1(*args) if cond else f2(*args)``.
+
+    Parameters
+    ----------
+    cond : array
+        The condition, expressed as a boolean array.
+    args : Array or tuple of Arrays
+        Argument(s) to `f1` (and `f2`). Must be broadcastable with `cond`.
+    f1 : callable
+        Elementwise function of `args`, returning a single array.
+        Where `cond` is True, output will be ``f1(arg0[cond], arg1[cond], ...)``.
+    f2 : callable, optional
+        Elementwise function of `args`, returning a single array.
+        Where `cond` is False, output will be ``f2(arg0[cond], arg1[cond], ...)``.
+        Mutually exclusive with `fill_value`.
+    fill_value : Array or scalar, optional
+        If provided, value with which to fill output array where `cond` is False.
+        It does not need to be scalar; it needs however to be broadcastable with
+        `cond` and `args`.
+        Mutually exclusive with `f2`. You must provide one or the other.
+    xp : array_namespace, optional
+        The standard-compatible namespace for `cond` and `args`. Default: infer.
+
+    Returns
+    -------
+    Array
+        An array with elements from the output of `f1` where `cond` is True and either
+        the output of `f2` or `fill_value` where `cond` is False. The returned array has
+        data type determined by type promotion rules between the output of `f1` and
+        either `fill_value` or the output of `f2`.
+
+    Notes
+    -----
+    Falls back to _lazywhere if xpx.apply_where is not available.
+
+    ``xp.where(cond, f1(*args), f2(*args))`` requires explicitly evaluating `f1` even
+    when `cond` is False, and `f2` when cond is True. This function evaluates each
+    function only for their matching condition, if the backend allows for it.
+
+    On Dask, `f1` and `f2` are applied to the individual chunks and should use functions
+    from the namespace of the chunks.
+
+    """
+    try:
+        import scipy._lib.array_api_extra as xpx
+
+        return xpx.apply_where(cond, args, f1, f2, fill_value=fill_value)
+    except (ImportError, AttributeError):
+        from scipy._lib._util import _lazywhere
+
+        return _lazywhere(cond, args, f1, fill_value, f2)

--- a/statsmodels/distributions/discrete.py
+++ b/statsmodels/distributions/discrete.py
@@ -1,24 +1,24 @@
-import numpy as np
+from statsmodels.compat.scipy import apply_where
 
-from scipy.stats import rv_discrete, poisson, nbinom
+import numpy as np
 from scipy.special import gammaln
-from scipy._lib._util import _lazywhere
+from scipy.stats import nbinom, poisson, rv_discrete
 
 from statsmodels.base.model import GenericLikelihoodModel
 
 
 class genpoisson_p_gen(rv_discrete):
-    '''Generalized Poisson distribution
-    '''
+    """Generalized Poisson distribution"""
+
     def _argcheck(self, mu, alpha, p):
-        return (mu >= 0) & (alpha==alpha) & (p > 0)
+        return (mu >= 0) & (alpha == alpha) & (p > 0)
 
     def _logpmf(self, x, mu, alpha, p):
-        mu_p = mu ** (p - 1.)
+        mu_p = mu ** (p - 1.0)
         a1 = np.maximum(np.nextafter(0, 1), 1 + alpha * mu_p)
-        a2 = np.maximum(np.nextafter(0, 1), mu + (a1 - 1.) * x)
-        logpmf_ = np.log(mu) + (x - 1.) * np.log(a2)
-        logpmf_ -=  x * np.log(a1) + gammaln(x + 1.) + a2 / a1
+        a2 = np.maximum(np.nextafter(0, 1), mu + (a1 - 1.0) * x)
+        logpmf_ = np.log(mu) + (x - 1.0) * np.log(a2)
+        logpmf_ -= x * np.log(a1) + gammaln(x + 1.0) + a2 / a1
         return logpmf_
 
     def _pmf(self, x, mu, alpha, p):
@@ -28,26 +28,27 @@ class genpoisson_p_gen(rv_discrete):
         return mu
 
     def var(self, mu, alpha, p):
-        dispersion_factor = (1 + alpha * mu**(p - 1))**2
+        dispersion_factor = (1 + alpha * mu ** (p - 1)) ** 2
         var = dispersion_factor * mu
         return var
 
 
-genpoisson_p = genpoisson_p_gen(name='genpoisson_p',
-                                longname='Generalized Poisson')
+genpoisson_p = genpoisson_p_gen(name="genpoisson_p", longname="Generalized Poisson")
 
 
 class zipoisson_gen(rv_discrete):
-    '''Zero Inflated Poisson distribution
-    '''
+    """Zero Inflated Poisson distribution"""
+
     def _argcheck(self, mu, w):
-        return (mu > 0) & (w >= 0) & (w<=1)
+        return (mu > 0) & (w >= 0) & (w <= 1)
 
     def _logpmf(self, x, mu, w):
-        return _lazywhere(x != 0, (x, mu, w),
-                          (lambda x, mu, w: np.log(1. - w) + x * np.log(mu) -
-                          gammaln(x + 1.) - mu),
-                          np.log(w + (1. - w) * np.exp(-mu)))
+        return apply_where(
+            x != 0,
+            (x, mu, w),
+            (lambda x, mu, w: np.log(1.0 - w) + x * np.log(mu) - gammaln(x + 1.0) - mu),
+            fill_value=np.log(w + (1.0 - w) * np.exp(-mu)),
+        )
 
     def _pmf(self, x, mu, w):
         return np.exp(self._logpmf(x, mu, w))
@@ -72,28 +73,32 @@ class zipoisson_gen(rv_discrete):
 
     def var(self, mu, w):
         dispersion_factor = 1 + w * mu
-        var = (dispersion_factor * self.mean(mu, w))
+        var = dispersion_factor * self.mean(mu, w)
         return var
 
     def _moment(self, n, mu, w):
         return (1 - w) * poisson.moment(n, mu)
 
 
-zipoisson = zipoisson_gen(name='zipoisson',
-                          longname='Zero Inflated Poisson')
+zipoisson = zipoisson_gen(name="zipoisson", longname="Zero Inflated Poisson")
+
 
 class zigeneralizedpoisson_gen(rv_discrete):
-    '''Zero Inflated Generalized Poisson distribution
-    '''
+    """Zero Inflated Generalized Poisson distribution"""
+
     def _argcheck(self, mu, alpha, p, w):
-        return (mu > 0) & (w >= 0) & (w<=1)
+        return (mu > 0) & (w >= 0) & (w <= 1)
 
     def _logpmf(self, x, mu, alpha, p, w):
-        return _lazywhere(x != 0, (x, mu, alpha, p, w),
-                          (lambda x, mu, alpha, p, w: np.log(1. - w) +
-                          genpoisson_p.logpmf(x, mu, alpha, p)),
-                          np.log(w + (1. - w) *
-                          genpoisson_p.pmf(x, mu, alpha, p)))
+        return apply_where(
+            x != 0,
+            (x, mu, alpha, p, w),
+            (
+                lambda x, mu, alpha, p, w: np.log(1.0 - w)
+                + genpoisson_p.logpmf(x, mu, alpha, p)
+            ),
+            fill_value=np.log(w + (1.0 - w) * genpoisson_p.pmf(x, mu, alpha, p)),
+        )
 
     def _pmf(self, x, mu, alpha, p, w):
         return np.exp(self._logpmf(x, mu, alpha, p, w))
@@ -103,29 +108,30 @@ class zigeneralizedpoisson_gen(rv_discrete):
 
     def var(self, mu, alpha, p, w):
         p = p - 1
-        dispersion_factor = (1 + alpha * mu ** p) ** 2 + w * mu
-        var = (dispersion_factor * self.mean(mu, alpha, p, w))
+        dispersion_factor = (1 + alpha * mu**p) ** 2 + w * mu
+        var = dispersion_factor * self.mean(mu, alpha, p, w)
         return var
 
 
 zigenpoisson = zigeneralizedpoisson_gen(
-    name='zigenpoisson',
-    longname='Zero Inflated Generalized Poisson')
+    name="zigenpoisson", longname="Zero Inflated Generalized Poisson"
+)
 
 
 class zinegativebinomial_gen(rv_discrete):
-    '''Zero Inflated Generalized Negative Binomial distribution
-    '''
+    """Zero Inflated Generalized Negative Binomial distribution"""
+
     def _argcheck(self, mu, alpha, p, w):
-        return (mu > 0) & (w >= 0) & (w<=1)
+        return (mu > 0) & (w >= 0) & (w <= 1)
 
     def _logpmf(self, x, mu, alpha, p, w):
         s, p = self.convert_params(mu, alpha, p)
-        return _lazywhere(x != 0, (x, s, p, w),
-                          (lambda x, s, p, w: np.log(1. - w) +
-                          nbinom.logpmf(x, s, p)),
-                          np.log(w + (1. - w) *
-                          nbinom.pmf(x, s, p)))
+        return apply_where(
+            x != 0,
+            (x, s, p, w),
+            (lambda x, s, p, w: np.log(1.0 - w) + nbinom.logpmf(x, s, p)),
+            fill_value=np.log(w + (1.0 - w) * nbinom.pmf(x, s, p)),
+        )
 
     def _pmf(self, x, mu, alpha, p, w):
         return np.exp(self._logpmf(x, mu, alpha, p, w))
@@ -153,7 +159,7 @@ class zinegativebinomial_gen(rv_discrete):
 
     def var(self, mu, alpha, p, w):
         dispersion_factor = 1 + alpha * mu ** (p - 1) + w * mu
-        var = (dispersion_factor * self.mean(mu, alpha, p, w))
+        var = dispersion_factor * self.mean(mu, alpha, p, w)
         return var
 
     def _moment(self, n, mu, alpha, p, w):
@@ -161,17 +167,19 @@ class zinegativebinomial_gen(rv_discrete):
         return (1 - w) * nbinom.moment(n, s, p)
 
     def convert_params(self, mu, alpha, p):
-        size = 1. / alpha * mu**(2-p)
+        size = 1.0 / alpha * mu ** (2 - p)
         prob = size / (size + mu)
         return (size, prob)
 
-zinegbin = zinegativebinomial_gen(name='zinegbin',
-    longname='Zero Inflated Generalized Negative Binomial')
+
+zinegbin = zinegativebinomial_gen(
+    name="zinegbin", longname="Zero Inflated Generalized Negative Binomial"
+)
 
 
 class truncatedpoisson_gen(rv_discrete):
-    '''Truncated Poisson discrete random variable
-    '''
+    """Truncated Poisson discrete random variable"""
+
     # TODO: need cdf, and rvs
 
     def _argcheck(self, mu, truncation):
@@ -195,18 +203,21 @@ class truncatedpoisson_gen(rv_discrete):
         loc = pmf < 1
         log_1_m_pmf[loc] = np.log(1 - pmf[loc])
         logpmf_ = poisson.logpmf(x, mu) - log_1_m_pmf
-        #logpmf_[x < truncation + 1] = - np.inf
+        # logpmf_[x < truncation + 1] = - np.inf
         return logpmf_
 
     def _pmf(self, x, mu, truncation):
         return np.exp(self._logpmf(x, mu, truncation))
 
-truncatedpoisson = truncatedpoisson_gen(name='truncatedpoisson',
-                                        longname='Truncated Poisson')
+
+truncatedpoisson = truncatedpoisson_gen(
+    name="truncatedpoisson", longname="Truncated Poisson"
+)
+
 
 class truncatednegbin_gen(rv_discrete):
-    '''Truncated Generalized Negative Binomial (NB-P) discrete random variable
-    '''
+    """Truncated Generalized Negative Binomial (NB-P) discrete random variable"""
+
     def _argcheck(self, mu, alpha, p, truncation):
         return (mu >= 0) & (truncation >= -1)
 
@@ -233,12 +244,15 @@ class truncatednegbin_gen(rv_discrete):
         return np.exp(self._logpmf(x, mu, alpha, p, truncation))
 
     def convert_params(self, mu, alpha, p):
-        size = 1. / alpha * mu**(2-p)
+        size = 1.0 / alpha * mu ** (2 - p)
         prob = size / (size + mu)
         return (size, prob)
 
-truncatednegbin = truncatednegbin_gen(name='truncatednegbin',
-    longname='Truncated Generalized Negative Binomial')
+
+truncatednegbin = truncatednegbin_gen(
+    name="truncatednegbin", longname="Truncated Generalized Negative Binomial"
+)
+
 
 class DiscretizedCount(rv_discrete):
     """Count distribution based on discretized distribution
@@ -330,9 +344,10 @@ class DiscretizedCount(rv_discrete):
         args, scale = self._unpack_args(args)
         if size is None:
             size = getattr(self, "_size", 1)
-        rv = np.trunc(self.distr.rvs(*args, scale=scale, size=size,
-                                     random_state=random_state) +
-                      self.d_offset)
+        rv = np.trunc(
+            self.distr.rvs(*args, scale=scale, size=size, random_state=random_state)
+            + self.d_offset
+        )
         return rv
 
     def _pmf(self, x, *args):
@@ -342,8 +357,7 @@ class DiscretizedCount(rv_discrete):
 
         args, scale = self._unpack_args(args)
 
-        p = (distr.sf(x, *args, scale=scale) -
-             distr.sf(x + 1, *args, scale=scale))
+        p = distr.sf(x, *args, scale=scale) - distr.sf(x + 1, *args, scale=scale)
         return p
 
     def _cdf(self, x, *args):
@@ -416,12 +430,13 @@ class DiscretizedModel(GenericLikelihoodModel):
     >>> probs = res.predict(which="probs", k_max=5)
 
     """
+
     def __init__(self, endog, exog=None, distr=None):
         if exog is not None:
             raise ValueError("exog is not supported")
 
         super().__init__(endog, exog, distr=distr)
-        self._init_keys.append('distr')
+        self._init_keys.append("distr")
         self.df_resid = len(endog) - distr.k_shapes
         self.df_model = 0
         self.k_extra = distr.k_shapes  # no constant subtracted
@@ -451,8 +466,7 @@ class DiscretizedModel(GenericLikelihoodModel):
             raise ValueError('only which="probs" is currently implemented')
 
     def get_distr(self, params):
-        """frozen distribution instance of the discrete distribution.
-        """
+        """frozen distribution instance of the discrete distribution."""
         args = params
         distr = self.distr(*args)
         return distr

--- a/statsmodels/genmod/_tweedie_compound_poisson.py
+++ b/statsmodels/genmod/_tweedie_compound_poisson.py
@@ -16,8 +16,10 @@ Jørgensen, B., Demétrio, C.G.B., Kristensen, E., Banta, G.T., Petersen, H.C.,
 Smyth G.K. and Jørgensen B. 2002. Fitting Tweedie's compound Poisson model to
     insurance claims data: dispersion modelling. ASTIN Bulletin 32: 143–157
 """
+
+from statsmodels.compat.scipy import apply_where
+
 import numpy as np
-from scipy._lib._util import _lazywhere
 from scipy.special import gammaln
 
 
@@ -31,9 +33,13 @@ def _alpha(p):
 
 def _logWj(y, j, p, phi):
     alpha = _alpha(p)
-    logz = (-alpha * np.log(y) + alpha * np.log(p - 1) - (1 - alpha) *
-            np.log(phi) - np.log(2 - p))
-    return (j * logz - gammaln(1 + j) - gammaln(-alpha * j))
+    logz = (
+        -alpha * np.log(y)
+        + alpha * np.log(p - 1)
+        - (1 - alpha) * np.log(phi)
+        - np.log(2 - p)
+    )
+    return j * logz - gammaln(1 + j) - gammaln(-alpha * j)
 
 
 def kappa(mu, p):
@@ -76,20 +82,21 @@ def density_otherwise(y, mu, p, phi):
 
 
 def series_density(y, mu, p, phi):
-    density = _lazywhere(np.array(y) > 0,
-                         (y, mu, p, phi),
-                         f=density_otherwise,
-                         f2=density_at_zero)
+    density = apply_where(
+        np.array(y) > 0, (y, mu, p, phi), f1=density_otherwise, f2=density_at_zero
+    )
     return density
 
 
-if __name__ == '__main__':
+if __name__ == "__main__":
     from scipy import stats
-    n = stats.poisson.rvs(.1, size=10000000)
-    y = stats.gamma.rvs(.1, scale=30000, size=10000000)
+
+    n = stats.poisson.rvs(0.1, size=10000000)
+    y = stats.gamma.rvs(0.1, scale=30000, size=10000000)
     y = n * y
     mu = stats.gamma.rvs(10, scale=30, size=10000000)
     import time
+
     t = time.time()
     out = series_density(y=y, mu=mu, p=1.5, phi=20)
-    print(f'That took {time.time() - t} seconds')
+    print(f"That took {time.time() - t} seconds")

